connectors/agentic-ai/src/main/java/io/camunda/connector/agenticai/aiagent/agent/DefaultAiAgentRequestHandler.java

```diff
@@ -0,0 +1,158 @@
+                Optional.ofNullable(requestData.memory().maxMessages())
+                    .orElse(DEFAULT_MAX_MEMORY_MESSAGES))
+            .chatMemoryStore(chatMemoryStore)
+            .build();
+
+    // check configured guardrails
+    checkGuardrails(requestData, agentContext);
+
+    // update memory with system + new user messages/tool call responses
+    addSystemPromptIfNecessary(chatMemory, requestData);
+    addUserMessagesFromRequest(agentContext, chatMemory, requestData);
+
+    // call LLM API with updates messages + resolved tool specifications
+    final var toolSpecifications =
+        toolCallingHandler.loadToolSpecifications(
+            context.getJobContext().getProcessDefinitionKey(),
+            requestData.tools().containerElementId());
+
+    final var chatRequest =
+        ChatRequest.builder()
+            .messages(chatMemory.messages())
+            .toolSpecifications(toolSpecifications)
+            .build();
+
+    final ChatResponse chatResponse = chatModel.chat(chatRequest);
+    final AiMessage aiMessage = chatResponse.aiMessage();
+    chatMemory.add(aiMessage);
+
+    // extract tool call requests from LLM response
+    final var toolsToCall = toolCallingHandler.extractToolsToCall(toolSpecifications, aiMessage);
+    final var nextAgentState =
+        !toolsToCall.isEmpty() ? AgentState.WAITING_FOR_TOOL_INPUT : AgentState.READY;
+
+    // update context
+    final var updatedContext =
+        chatMemoryStore
+            .storeToAgentContext(agentContext)
+            .withState(nextAgentState)
+            .withMetrics(
+                agentContext
+                    .metrics()
+                    .incrementModelCalls(1)
+                    .incrementTokenUsage(AgentMetrics.TokenUsage.from(chatResponse.tokenUsage())));
+
+    return new AgentResponse(updatedContext, updatedContext.memory().getLast(), toolsToCall);
+  }
+
+  private void checkGuardrails(
+      AgentRequest.AgentRequestData requestData, AgentContext agentContext) {
+    // naive guardrail implementation
+    final int maxModelCalls =
+        Optional.ofNullable(requestData.guardrails().maxModelCalls())
+            .orElse(DEFAULT_MAX_MODEL_CALLS);
+    if (agentContext.metrics().modelCalls() >= maxModelCalls) {
+      throw new ConnectorException("Maximum number of model calls reached");
+    }
+  }
+
+  private void addSystemPromptIfNecessary(
+      ChatMemory chatMemory, AgentRequest.AgentRequestData requestData) {
+    // add system prompt if this is the first request
+    if (chatMemory.messages().isEmpty()) {
+      chatMemory.add(promptFromConfiguration(requestData.systemPrompt()).toSystemMessage());
+    }
+  }
+
+  private void addUserMessagesFromRequest(
+      AgentContext agentContext, ChatMemory chatMemory, AgentRequest.AgentRequestData requestData) {
+    if (agentContext.isInState(AgentState.WAITING_FOR_TOOL_INPUT)) {
+      final var toolCallResults = requestData.tools().toolCallResults();
```

Comment:
`toolCallResults` is the multi-instance output collection which is injected as part of the connector config:

![image](https://github.com/user-attachments/assets/f527221e-949c-4a78-be2d-7bf02eccb6d5)
![image](https://github.com/user-attachments/assets/c3ff64b1-d524-461a-825a-0c3607e3d91c)